awscli/errorhandler.py

```diff
@@ -11,68 +11,110 @@
-                http_status_code=http_response.status_code)
-
-    def _get_error_code_and_message(self, response):
-        code = 'Unknown'
-        message = 'Unknown'
-        if 'Error' in response:
-            error = response['Error']
-            return error.get('Code', code), error.get('Message', message)
-        return (code, message)
+class BaseExceptionHandler:
+    def handle_exception(self, exception, stdout, stderr):
+        raise NotImplementedError('handle_exception')
+
+
+class FilteredExceptionHandler(BaseExceptionHandler):
+    EXCEPTIONS_TO_HANDLE = ()
+    MESSAGE = '%s'
+
+    def handle_exception(self, exception, stdout, stderr):
+        if issubclass(exception.__class__, self.EXCEPTIONS_TO_HANDLE):
+            return_val = self._do_handle_exception(exception, stdout, stderr)
+            if return_val is not None:
+                return return_val
+
+    def _do_handle_exception(self, exception, stdout, stderr):
+        stderr.write("\n")
+        stderr.write(self.MESSAGE % exception)
+        stderr.write("\n")
+        return self.RC
+
+
+class ParamValidationErrorsHandler(FilteredExceptionHandler):
+    EXCEPTIONS_TO_HANDLE = (
+        ParamError, ParamSyntaxError, ArgParseException,
+        ParamValidationError, BotocoreParamValidationError,
+        UnknownArgumentError
+    )
+    RC = PARAM_VALIDATION_ERROR_RC
+
+
+class ClientErrorHandler(FilteredExceptionHandler):
+    EXCEPTIONS_TO_HANDLE = ClientError
+    RC = CLIENT_ERROR_RC
+
+
+class ConfigurationErrorHandler(FilteredExceptionHandler):
+    EXCEPTIONS_TO_HANDLE = ConfigurationError
+    RC = CONFIGURATION_ERROR_RC
+
+
+class NoRegionErrorHandler(FilteredExceptionHandler):
+    EXCEPTIONS_TO_HANDLE = NoRegionError
+    RC = CONFIGURATION_ERROR_RC
+    MESSAGE = '%s You can also configure your region by running "aws configure".'
+
+
+class NoCredentialsErrorHandler(FilteredExceptionHandler):
+    EXCEPTIONS_TO_HANDLE = NoCredentialsError
+    RC = CONFIGURATION_ERROR_RC
+    MESSAGE = '%s. You can configure credentials by running "aws configure".'
+
+
+class UnknownArgumentErrorHandler(FilteredExceptionHandler):
+    EXCEPTIONS_TO_HANDLE = UnknownArgumentError
+    RC = PARAM_VALIDATION_ERROR_RC
+
+    def _do_handle_exception(self, exception, stdout, stderr):
+        stderr.write("\n")
+        stderr.write(f'usage: {USAGE}\n{exception}\n')
+        stderr.write("\n")
+        return self.RC
+
+
+class InterruptExceptionHandler(FilteredExceptionHandler):
+    EXCEPTIONS_TO_HANDLE = KeyboardInterrupt
+    RC = 128 + signal.SIGINT
+
+
+class GeneralExceptionHandler(FilteredExceptionHandler):
+    EXCEPTIONS_TO_HANDLE = Exception
+    RC = GENERAL_ERROR_RC
+
+
+class ChainedExceptionHandler(BaseExceptionHandler):
+    def __init__(self, exception_handlers=None):
+        self._exception_handlers = exception_handlers
+        if self._exception_handlers is None:
+            self._exception_handlers = [GeneralExceptionHandler]
```

Comment:
Let's just make `exception_handlers` a required parameter here and use the list provided instead of have a default handler. If we want to have the `GeneralExceptionHandler` be the only exception handler, I'd imagine we just use that directly instead of going through a chained handler.